pkg/monitor/resourcewatch/storage/git_store.go

```diff
@@ -103,34 +147,68 @@ func (s *GitStorage) handle(gvr schema.GroupVersionResource, oldObj, obj *unstru
 		modifyingUser = err.Error()
 	}
 
-	// either the golang git library or git itself doesn't properly handle threading, so we have to lock here.
-	s.Lock()
-	defer s.Unlock()
+	// ignore error, we've already reported and we're not doing anything else.
+	_ = wait.PollImmediate(1*time.Second, 15*time.Second, func() (bool, error) {
+		// either the golang git library or git itself doesn't properly handle threading, so we have to lock here.
+		// take the lock inside the poll so that retries don't keep the lock for a long time.
+		s.Lock()
+		defer s.Unlock()
 
-	switch {
-	case operation == gitOpAdded:
-		if err := s.commitAdd(filePath, modifyingUser, ocCommand); err != nil {
-			klog.Error(err)
-		}
-	case operation == gitOpModified:
-		if err := s.commitModify(filePath, modifyingUser, ocCommand); err != nil {
-			klog.Error(err)
+		switch {
+		case operation == gitOpAdded:
+			if err := s.commitAdd(filePath, modifyingUser, ocCommand); err != nil {
+				klog.Error(err)
+				return false, nil
+			}
+		case operation == gitOpModified:
+			if err := s.commitModify(filePath, modifyingUser, ocCommand); err != nil {
+				klog.Error(err)
+				return false, nil
+			}
+		default:
+			klog.Error("unhandled case")
+
+			return true, nil
 		}
-	default:
-		klog.Error("unhandled case")
-
-	}
+		return true, nil
+	})
 }
 
 func (s *GitStorage) OnAdd(gvr schema.GroupVersionResource, obj interface{}) {
 	objUnstructured := obj.(*unstructured.Unstructured)
-	s.handle(gvr, nil, objUnstructured, false)
+
+	// serialize updates to individual files
+	key := fmt.Sprintf("%s/%s/%s/%s/%s", gvr.Group, gvr.Version, gvr.Resource, objUnstructured.GetNamespace(), objUnstructured.GetName())
+	if err := s.currentlyRecording.waitUntilAvailable(key); err != nil {
```

Comment:
I don't remember the details of dynamic informers. Will OnAdd/OnUpdate/OnDelete for the same key be called from different threads?

Reply:
>I don't remember the details of dynamic informers. Will OnAdd/OnUpdate/OnDelete for the same key be called from different threads?

For each informer, OnAdd/OnUpdate/OnDelete are all called serially